### ISSUE TITLE
DrugCLIP training procedure

### ISSUE THREAD

**Issue by hzau96yhz** [bowen-gao/DrugCLIP#1]
Hi！Your work is very inspiring！I have a question about the training procedure of DrugCLIP.
In the CLIP model, the image-text pair data in a certain dataset is used in the pre-training process. In this paper, figure 1 represents an example of the training process of the drugclip model. The paper describes "**At each training iteration, sampled 3D molecule and 3D pocket representations are learned with a contrastive objective**".
As I understand it, for each molecular-protein pair data, where each molecule has a different conformation, each protein is further augmented to get multiple proteins.
So how to interpret the similarity matrix on the right-hand side of Figure 1? The diagonal elements in the similarity matrix do not appear to be every pair of molecular-protein data in the dataset. For a molecular-protein pair data in the dataset, it seems that different conformations of the molecule and different similar pockets of the protein should both be positive examples, right? I do not know if this understanding is correct, I hope to get your answer!

**Comment by bowen-gao** [comment]
Yes, your understanding is correct

**Comment by hzau96yhz** [comment]
Hi！Sorry to bother you again！
![Snipaste_2023-12-06_20-56-06](https://github.com/bowen-gao/DrugCLIP/assets/53745234/9ade715d-ad02-4211-bf96-c5317607f4a6)
The paper describes **"At each training iteration, sampled 3D molecule and 3D pocket representations are learned with a contrastive objective"**. As shown in the figure above, for each pair of small molecular-protein data, after generating different small molecule conformations and homologous proteins, one of the small molecules and proteins is randomly sampled as a positive sample pair. 

However, the negative samples in this case also seem to represent one of the different molecular conformations and one of the different homologous proteins. It seems that this pair is also from the original molecular-protein pair data and cannot be counted as a negative sample. 

Hope to get your explanation from the author.

**Comment by bowen-gao** [comment]
We utilize 'in-batch negatives' in our approach. In this method, the negative samples are derived from pockets or molecules that are part of other positive pairs within the same batch. It's important to clarify that data produced through augmentation techniques are still categorized as positive pairs and are not used as negatives

**Comment by bowen-gao** [comment]
I see what you are asking. Molecule conformations are sampled across
different epochs. For homo augs, we add them directly into the dataset.
Actually, we use a masking technique to prevent pockets that originate from
the same source from being incorrectly identified as negative samples
within the same batch. This masking is also applied to molecules to avoid
similar issues. Hope this answers your question！

**Comment by bowen-gao** [comment]
In the original dataset, there are N pairs of (pocket, ligand). After introducing homologs at a rate of three per pair, the total count of pairs increases to 4N, comprising the original N pairs and an additional 3N pairs created from the homologs. During each batch, we randomly select B pairs from this expanded set of 4N pairs. Furthermore, for each molecule in the batch, we sample one conformation out of a possible K conformations. It's important to note that the selected conformation for any given molecule may vary in different epochs.

If you have further questions, you can find me at billgao0111@gmail.com to have a chat. It will be more convenient.
